File: samples/vdsp_op/flip/flip.py
import vaststreamx as vsx
import cv2
import argparse
import os
import sys

current_file_path = os.path.dirname(os.path.abspath(__file__))
common_path = os.path.join(current_file_path, "../../..")
sys.path.append(common_path)
import common.utils as utils


attr = vsx.AttrKey


def argument_parser():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "-d",
        "--device_id",
        default=0,
        type=int,
        help="device id to run",
    )
    parser.add_argument(
        "--input_file",
        default="data/images/dog.jpg",
        help="input image",
    )
    parser.add_argument(
        "--output_file",
        default="flip_resul.jpg",
        help="output image",
    )
    parser.add_argument(
        "--flip_type",
        default="x",
        help="flip type x or y",
    )
    args = parser.parse_args()
    return args


if __name__ == "__main__":
    args = argument_parser()
    vsx.set_device(args.device_id)
    cv_bgr888 = cv2.imread(args.input_file)
    assert cv_bgr888 is not None, f"Failed to read: {args.input_file}"
    height, width = cv_bgr888.shape[:2]
    cv_nv12 = utils.cv_bgr888_to_nv12(cv_bgr888)
    vsx_image = vsx.create_image(
        cv_nv12,
        vsx.ImageFormat.YUV_NV12,
        width,
        height,
        args.device_id,
    )
    # Build operator
    op = vsx.BuildInOperator(vsx.OpType.SINGLE_OP_FLIP)

    if args.flip_type == "x" or args.flip_type == "X":
        flip_type = vsx.ImageFlipType.FLIP_TYPE_X_AXIS
    elif args.flip_type == "y" or args.flip_type == "Y":
        flip_type = vsx.ImageFlipType.FLIP_TYPE_Y_AXIS
    else:
        print(f'Unsupport flip type: "{args.flip_type}", only support x or y')
        exit(-1)

    op.set_attribute(
        {
            attr.IIMAGE_FORMAT: vsx.ImageType.YUV_NV12,
            attr.IIMAGE_WIDTH: vsx_image.width,
            attr.IIMAGE_HEIGHT: vsx_image.height,
            attr.IIMAGE_WIDTH_PITCH: vsx_image.width,
            attr.IIMAGE_HEIGHT_PITCH: vsx_image.height,
            attr.OIMAGE_WIDTH: vsx_image.width,
            attr.OIMAGE_HEIGHT: vsx_image.height,
            attr.OIMAGE_WIDTH_PITCH: vsx_image.width,
            attr.OIMAGE_HEIGHT_PITCH: vsx_image.height,
            attr.DIRECTION: flip_type,
        }
    )
    # run
    output_list = op.execute(
        [vsx_image],
        [(vsx.ImageFormat.YUV_NV12, vsx_image.width, vsx_image.height)],
    )

    out_images = [vsx.as_numpy(output).squeeze() for output in output_list]
    out_bgr888 = utils.cv_nv12_to_bgr888(out_images[0])
    cv2.imwrite(args.output_file, out_bgr888)

    print(f"save result to {args.output_file}")

File: samples/vdsp_op/flip/flip.cpp

/*
 * Copyright (C) 2025 Vastai-tech Company.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
#include <chrono>
#include <thread>

#include "common/cmdline.hpp"
#include "common/utils.hpp"
#include "opencv2/opencv.hpp"
#include "vaststreamx/vaststreamx.h"

cmdline::parser ArgumentParser(int argc, char** argv) {
  cmdline::parser args;
  args.add<uint32_t>("device_id", 'd', "device id to run", false, 0);
  args.add<std::string>("input_file", '\0', "input_image", false,
                        "../data/images/dog.jpg");
  args.add<std::string>("output_file", '\0', "output image", false,
                        "flip_result.jpg");
  args.add<std::string>("flip_type", '\0', "flip type x or y", false, "x");
  args.parse_check(argc, argv);
  return args;
}

int main(int argc, char* argv[]) {
  auto args = ArgumentParser(argc, argv);
  uint32_t device_id = args.get<uint32_t>("device_id");
  auto input_file = args.get<std::string>("input_file");
  auto flip_type_str = args.get<std::string>("flip_type");

  vsx::ImageFlipType flip_type;

  if (flip_type_str == "x" || flip_type_str == "X") {
    flip_type = vsx::ImageFlipType::kFLIP_TYPE_X_AXIS;
  } else if (flip_type_str == "y" || flip_type_str == "Y") {
    flip_type = vsx::ImageFlipType::kFLIP_TYPE_Y_AXIS;
  } else {
    std::cout << "Unsupport flip type: \"" << flip_type_str
              << "\", only support x or y\n";
    return -1;
  }
  vsx::SetDevice(device_id);

  vsx::Image image;
  if (vsx::MakeVsxImage(input_file, image, vsx::YUV_NV12) != 0) {
    std::cout << "Read input image failed: " << input_file << std::endl;
  }
  int iimage_width = image.Width();
  int iimage_height = image.Height();
  auto image_format = image.Format();

  auto op = vsx::BuildInOperator(vsx::BuildInOperatorType::kSINGLE_OP_FLIP);

  op.SetAttribute<vsx::AttrKey::kIimageWidth>(iimage_width);
  op.SetAttribute<vsx::AttrKey::kIimageHeight>(iimage_height);
  op.SetAttribute<vsx::AttrKey::kIimageWidthPitch>(iimage_width);
  op.SetAttribute<vsx::AttrKey::kIimageHeightPitch>(iimage_height);
  op.SetAttribute<vsx::AttrKey::kOimageWidth>(iimage_width);
  op.SetAttribute<vsx::AttrKey::kOimageHeight>(iimage_height);
  op.SetAttribute<vsx::AttrKey::kOimageWidthPitch>(iimage_width);
  op.SetAttribute<vsx::AttrKey::kOimageHeightPitch>(iimage_height);
  op.SetAttribute<vsx::AttrKey::kIimageFormat>(
      vsx::BuildInOperatorAttrImageType::kYUV_NV12);
  op.SetAttribute<vsx::AttrKey::kDirection>(flip_type);

  // execute

  // copy input to device memory
  auto input_vacc = vsx::Image(image_format, iimage_width, iimage_height,
                               vsx::Context::VACC(device_id));
  input_vacc.CopyFrom(image);

  // make device memory image for output
  auto output_vacc = vsx::Image(image_format, iimage_width, iimage_height,
                                vsx::Context::VACC(device_id));
  std::vector<vsx::Image> outputs{output_vacc};
  // buildin operator execute
  op.Execute({input_vacc}, outputs);
  // copy result to cpu memory
  vsx::Image output_cpu = vsx::Image(output_vacc.Format(), output_vacc.Width(),
                                     output_vacc.Height());
  output_cpu.CopyFrom(output_vacc);

  cv::Mat out_mat;
  vsx::ConvertVsxImageToCvMatBgrPacked(output_cpu, out_mat);
  cv::imwrite(args.get<std::string>("output_file"), out_mat);
  std::cout << "save result to : " << args.get<std::string>("output_file")
            << std::endl;
  return 0;
}
